File: src/services/share.ts
import { invoke } from "@tauri-apps/api/core";

export interface UserOwner {
    id: string;
    nickname: string;
    email: string;
    createdAt: string;
}

export interface ShareItem {
    id: string;
    name: string;
    expired: boolean;
    source_type: number;
    unlocked: boolean;
    url: string;
    visited: number;
    created_at: string;
    owner: UserOwner;
}

export interface Pagination {
    total: number;
    page: number;
    pageSize: number;
}

export interface GetSharesAck {
    shares: ShareItem[];
    pagination: Pagination;
}


export interface ShareOwner {
    id: string;
    email: string;
    nickname: string;
    created_at: Date | string;
}

export interface ShareInfo {
    id: string;
    name: string;
    visited: number;
    unlocked: boolean;
    source_type: number;
    owner: ShareOwner;
    created_at: Date | string;
    expired: boolean;
    url: string;
    source_uri: string;
}


export const deleteShare = async (id: string) => {
    try {
        const response = await invoke<boolean>('delete_share', { id });
        return response;
    } catch (error) {
        console.error('Failed to delete share:', id, error);
        throw error;
    }
}


export const updateShare = async (id: string, downloads: number, expire: number, uri: string) => {
    try {
        const response = await invoke<string>('update_share', { id, downloads, expire, uri });
        return response;
    } catch (error) {
        console.error('Failed to delete share:', id, error);
        throw error;
    }
}

export const getShareInfo = async (id: string, ownerExtended: boolean) => {
    try {
        const response = await invoke<ShareInfo>('get_share_info', { id, ownerExtended });
        return response;
    } catch (error) {
        console.error('Failed to get share info:', id, error);
        throw error;
    }
}

File: src-tauri/src/proto/share.rs
use ::serde::{Deserialize, Serialize};

use super::storage::Pagination;

#[derive(Debug, Serialize, Deserialize)]
pub struct Share {
    pub id: String,
    pub name: String,
    pub expired: bool,
    pub source_type: u8,
    pub unlocked: bool,
    pub url: String,
    pub visited: u64,
    pub created_at: String,
    pub owner: UserOwner,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UserOwner {
    pub id: String,
    pub nickname: String,
    pub email: String,
    pub created_at: String,
}
#[derive(Debug, Serialize, Deserialize)]
pub struct GetSharesAck {
    pub shares: Vec<Share>,
    pub pagination: Pagination,
}
#[derive(Debug, Serialize, Deserialize)]
pub struct GetSharesReq {
    pub page_size: u64,
    pub order_direction: String,
}
#[derive(Debug, Serialize, Deserialize)]
pub struct ShareOwner {
    pub id: String,
    pub email: String,
    pub nickname: String,
    pub created_at: String,
}
#[derive(Debug, Serialize, Deserialize)]
pub struct ShareInfo {
    pub id: String,
    pub name: String,
    pub visited: i32,
    pub unlocked: bool,
    pub source_type: i32,
    pub owner: Option<ShareOwner>,
    pub created_at: String,
    pub expired: bool,
    pub url: String,
    pub source_uri: String,
}
